include/taskolib/Timeout.h

```diff
@@ -0,0 +1,183 @@
+    template <typename Rep, typename Period>
+    constexpr
+    Timeout(std::chrono::duration<Rep, Period> duration)
+    {
+        if (duration < duration.zero())
+            throw Error("Negative timeout");
+
+        if (duration >= infinite_duration)
+            timeout_ = infinite_duration;
+        else
+            timeout_ = std::chrono::round<Duration>(duration);
+    }
+
+    /**
+     * Construct a Timeout from a floating-point value (only explicitly).
+     *
+     * \param seconds  The timeout in seconds
+     */
+    template <typename FloatT,
+              std::enable_if_t<std::is_floating_point<FloatT>::value, bool> = true>
+    explicit constexpr
+    Timeout(FloatT seconds)
+    {
+        if (seconds < FloatT{ 0.0 })
+            throw Error("Negative timeout");
+
+        const auto desired_duration = std::chrono::duration<FloatT>{ seconds };
+
+        if (not std::isfinite(seconds) || desired_duration >= infinite_duration)
+            timeout_ = infinite_duration;
+        else
+            timeout_ = std::chrono::round<Duration>(desired_duration);
+    }
+
+    /// A constant to use for "infinite" timeout durations
+    static constexpr Timeout infinity() { return Timeout{ infinite_duration }; };
+
+    /// Determine if the timeout has a finite duration.
+    bool is_finite() const noexcept
+    {
+        return timeout_ < infinite_duration;
+    }
+
+    /**
+     * Determine if the timeout has an infinite duration (in other words, if there is no
+     * timeout).
+     */
+    bool is_infinite() const noexcept
+    {
+        return timeout_ >= infinite_duration;
+    }
+
+    /// Determine if two timeouts are equal.
+    friend constexpr bool operator==(Timeout lhs, Timeout rhs) { return lhs.timeout_ == rhs.timeout_; }
+
+    /// Determine if two timeouts are not equal.
+    friend constexpr bool operator!=(Timeout lhs, Timeout rhs) { return lhs.timeout_ != rhs.timeout_; }
+
+    /// Determine if the left timeout is longer than the right one.
+    friend constexpr bool operator> (Timeout lhs, Timeout rhs) { return lhs.timeout_ >  rhs.timeout_; }
+
+    /// Determine if the left timeout is shorter than the right one.
+    friend constexpr bool operator< (Timeout lhs, Timeout rhs) { return lhs.timeout_ <  rhs.timeout_; }
+
+    /// Determine if the left timeout is longer than or equal to the right one.
+    friend constexpr bool operator>=(Timeout lhs, Timeout rhs) { return lhs.timeout_ >= rhs.timeout_; }
+
+    /// Determine if the left timeout is shorter than or equal to the right one.
+    friend constexpr bool operator<=(Timeout lhs, Timeout rhs) { return lhs.timeout_ <= rhs.timeout_; }
+
+    /**
+     * Implicitly cast the timeout to a floating-point type.
+     * Infinity is returned for an infinite timeout.
+     */
+    template <typename FloatT,
+              std::enable_if_t<std::is_floating_point<FloatT>::value, bool> = true>
+    operator FloatT() const
```

Comment:
This is working?
I thought the template type deduction works only on parameters. :grimacing:

Reply:
No, this is working. I always copy the example from [cppreference](https://en.cppreference.com/w/cpp/types/enable_if) (the one under "/* RIGHT */").

Reply:
Yes. Still :grimacing: Done to much other languages lately.

Reply:
Well, we can agree that `enable_if` is ugly as hell. 😈